os/drivers/pm/pm.c

```diff
@@ -0,0 +1,154 @@
+
+/****************************************************************************
+ * Function Prototypes
+ ****************************************************************************/
+
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg);
+static ssize_t pm_read(FAR struct file *filep, FAR char *buffer, size_t len);
+static ssize_t pm_write(FAR struct file *filep, FAR const char *buffer, size_t len);
+
+/****************************************************************************
+ * Data
+ ****************************************************************************/
+static const struct file_operations pm_fops = {
+	0,                          /* open */
+	0,                          /* close */
+	pm_read,                    /* read */
+	pm_write,                   /* write */
+	0,                          /* seek */
+	pm_ioctl                    /* ioctl */
+#ifndef CONFIG_DISABLE_POLL
+	, 0                         /* poll */
+#endif
+};
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static ssize_t pm_read(FAR struct file *filep, FAR char *buffer, size_t len)
+{
+	return 0;
+}
+
+static ssize_t pm_write(FAR struct file *filep, FAR const char *buffer, size_t len)
+{
+	return 0;
+}
+
+
+/************************************************************************************
+ * Name: pm_ioctl
+ *
+ * Command description:
+ *   PMIOC_LOCK - for locking PM transition (both permanent and timer)
+ *   PMIOC_UNLOCK - for unlocking PM transiton
+ *   PMIOC_TIMER_CREATE - to create a wakeup timer
+ *   PMIOC_TIMER_SET - to set a wakeup timer
+ *   PMIOC_TIMER_STOP - to stop a specific wakeup timer
+ *   PMIOC_TIMER_DELETE - to delete a specific wakeup timer and make it available for other apps
+ *   PMIOC_TIMETICK - for getting current time tick
+ *   PMIOC_TUNEFREQ - for changing the operating frequency of the core to save power
+ * 
+ * Arguments:
+ *   filep is ioctl fd, cmd is required command, arg is required argument for
+ *   the command. 
+ *   for LOCK, arg should be a long unsigned type.
+ *   for TIMER_CREATE, arg should be a bool value.
+ *   for TIMER_SET, arg should be a pointer to struct pm_timer_header type.
+ *   for TIMER_STOP and TIMER_DELETE, arg should be a int (id of timer).
+ *   for TIMETICK , arg should be a pointer to a size_t type.
+ *   for TUNEFREQ, arg should be an int type.
+ *
+ * Description:
+ *   This api can be used to perform PM operation.
+ *
+ ************************************************************************************/
+static int pm_ioctl(FAR struct file *filep, int cmd, unsigned long arg)
+{
+	int ret = -ENOSYS;
+    
+	/* Handle built-in ioctl commands */
+	switch (cmd) {
+    	case PMIOC_LOCK:
+        	pm_stay(PM_IDLE_DOMAIN, PM_NORMAL);
+        	pmvdbg("State locked!\n");
+        	if (arg > 0) {
+            		pm_set_timer(PM_LOCK_TIMER, arg);
```

Comment:
Because I think it is better separate the pm lock and pm timer within the code entirely. 
These two operations are completely distinct from one another. 
the pm timer use just pm_lock.

Reply:
done